docs/internals/task_lifecycle.md

```diff
@@ -0,0 +1,149 @@
+
+Now the task can resume execution, and operators can start processing fresh input data. This is the place where the task-specific `run()` mehtod is called. This method will 
+run until either there is no more input data (finite stream), or the task is cancelled (manually or not). Here is where the operator specific `processElement()` and 
+`processWatermark()` methods are called.
+
+In the case of running till completion, *i.e.* there is no more input data to process, after exiting from the `run()` method, the task enters its shutdown process. 
+Initially, the timer service stops registering any new timers (e.g. from fired timers that are being executed), clears all the not yet started timers, and awaits the 
+completion of currently executing ones. Then the `closeAllOperators()` tries to gracefully close the operators involved in the computation by calling the `close()` method 
+of each one of them. Then any buffered output data is flushed so that they can be processed by the downstream tasks, and finally the task tries to clear all the resources 
+held by the operators by calling the `dispose()` method of each one of them. When opening the different operators, we mentioned that the order is from the last to the first. 
+Closing happens the other way round, from first to last.
+
+Finally, when all operators have been closed and all their resources freed, the task shuts down its timer service, performs its task-specific cleanup, *e.g.* clean all its 
+internal buffers, and then performs its generic task clean up which consists of closing all its output channels and cleaning any output buffers.
+
+**Checkpoints:** Previously we saw that during `initializeState()`, and in case of recovering from a failure, the task and all its operators and functions retrieve the 
+state that was persisted to stable storage during the last successful checkpoint before the failure. Checkpoints in Flink are performed periodically based on a user-specified 
+interval, and are performed by a different thread than that of the main task thread. This is the reason the are not included in the main phases of the task lifecycle. In a 
+nutshell, special elements called `CheckpointBarriers` are injected periodically by the source tasks of a job in the stream of input data, and travel with the actual data from 
+source to sink. A source task injects these barriers after it is in running mode, and assuming that the CheckpointCoordinator is also running. Whenever a task receives such a 
+barrier, it schedules a task to be performed by the checkpoint thread which calls the `snapshotState()` of the operators in the task. Input data can still be received by the 
+task while the checkpoint is being performed, but they are buffered and only emitted downstream after the checkpoint is successfully completed. For details on the principles 
```

Comment:
"but they are buffered ..." -> "but they are buffered and only *processed and* emitted downstream..."